text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+
+# Motivation
+[motivation]: #motivation
+
+Today in Rust you can use `as` casts on *field-less* `enum`s to get their discriminants,
+but as soon as any variant has fields, that's no longer available.
+
+Rust 1.66 stabilized custom discriminants on variants with fields, but as
+[the release post][rust 1.66 blog] said,
+
+> Rust provides no language-level way to access the raw discriminant of an enum with fields.
+> Instead, currently unsafe code must be used to inspect the discriminant of an enum with fields.
+
+[rust 1.66 blog]: https://blog.rust-lang.org/2022/12/15/Rust-1.66.0.html#explicit-discriminants-on-enums-with-fields
+
+As a result, the [documentation for `mem::Discriminant`][discriminant docs] has a section
+about how to write that `unsafe` code, and a bunch of warnings about the different
+*incorrect* ways that must not be used.
+
+[discriminant docs]: https://doc.rust-lang.org/std/mem/fn.discriminant.html#accessing-the-numeric-value-of-the-discriminant
+
+It's technically [possible](https://github.com/rust-lang/rust/pull/106418#issuecomment-1700399884)
+to write a clever enough safe `match` that compiles down to a no-op in order to get at the discriminant,
+but doing so is annoying and fragile.
+
+And accessing the discriminant is quite useful in various places, so it'd be nice for it to be easy.
+
+For example, `#[derive(PartialOrd)]` on an `enum` today uses internal compiler magic to look at discriminants.
+It would be nice for other derives in the ecosystem -- there's a whole bunch of things on `enum`s --
+to be able to look at the discriminants directly too.
+
+With this RFC, the built-in derives and third-party derives can both use the same stable feature
+to implement `PartialOrd::parial_cmp` for the cases where the arguments have different discriminants.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+[Rust 1.66][rust 1.66 blog] stabilized custom discriminants on enum variants,
+but didn't give a nice way to actually read them.
+
+In this release, you can use **`.enum#discriminant`** to read them.
+
+For example, if you have the following enum,
+
+```rust
+#[repr(u8)]
+enum Enum {
+    Unit = 7,
+    Tuple(bool) = 13,
+    Struct { a: i8 } = 42,
+}
+```
+
+Then the following examples pass:
+
+```rust
+let a = Enum::Unit;
+assert_eq!(a.enum#discriminant, 7);
+let b = Enum::Tuple(true);
+assert_eq!(b.enum#discriminant, 13);
+let c = Enum::Struct { a: 1 };
+assert_eq!(c.enum#discriminant, 42);
+```
+
+That's entirely safe code, and the value comes out as the type from the `repr`,
+avoiding the change to accidentally use a mismatched type.
+
+To avoid making implicit semver promises, this is only available for `enum`s
+that are defined in the current module.  If you want to expose it to others,
+feel free to define a method like
+
+```rust
+impl Enum {
+	pub fn discriminant(&self) -> u8 {
+		self.enum#discriminant
+	}
+}
+```
```

Comment:
This solution is not generally tenable for `repr(C)` enums, whose discriminant type is platform-dependent. We would also need some way to reflect on the type of the discriminant.

Reply:
Perhaps? Let's dig.

RFC2195 [says](https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md#guide-level-explanation:~:text=%23%5Brepr(C)%5D%20currently%20just%20tells%20Rust%20to%20try%20to%20pick%20whatever%20integer%20type%20that%20a%20C%20compiler%20for%20the%20target%20platform%20would%20use%20for%20an%20enum.):

> `#[repr(C)]` currently just tells Rust to try to pick whatever integer type that a C compiler for the target platform would use for an enum.

And the Rust Unsafe Code Guidelines [say](https://rust-lang.github.io/unsafe-code-guidelines/layout/enums.html#layout-of-a-fieldless-enum:~:text=The%20%23%5Brepr(C)%5D%20annotation%20is%20equivalent%2C%20but%20it%20selects%20the%20same%20size%20as%20the%20C%20compiler%20would%20use%20for%20the%20given%20target%20for%20an%20equivalent%20C%2Denum%20declaration.):

> When a `#[repr(Int)]`-style annotation is attached to a fieldless enum (one without any data for its variants), it will cause the enum to be represented as a simple integer of the specified size `Int`. [...] The `#[repr(C)]` annotation is equivalent, but it selects the same size as the C compiler would use for the given target for an equivalent C-enum declaration.

I'm not an expert in C. From a brief search, the relevant text in the C99 standard appears to be [here](https://port70.net/~nsz/c/c99/n1256.html#6.7.2.2p2):

> ###### Constraints
> The expression that defines the value of an enumeration constant shall be an integer constant expression that has a value representable as an int.

I *think* this guarantees that `c_int` is a suitable logical type of the discriminant, but it appears to leave open the possibility that the type used for the tag in C might be different. This is problematic for Rust, because it's commonly assumed that `repr(C)` enums have the same tag and discriminant types and values.

Reply:
If you don't care about privacy / stability / nightly etc there is already the [std::marker::DiscriminantKind](https://doc.rust-lang.org/std/marker/trait.DiscriminantKind.html) trait to extract the `Discriminant` type.

However, currently Rust always resolve this to `isize`, which is probably wrong?

```rust
#![feature(discriminant_kind)]

use std::{marker::DiscriminantKind, mem::size_of};

#[repr(C)]
pub enum E {
    A,
    B,
}

const _: [(); 4] = [(); size_of::<E>()]; // size of E is same as c_int
const _: [(); 8] = [(); size_of::<<E as DiscriminantKind>::Discriminant>()]; // size of discriminant is same as isize
```